standardization/src/main/scala/za/co/absa/enceladus/standardization/interpreter/stages/TypeParser.scala

```diff
@@ -15,185 +15,364 @@
+  val isArrayElement: Boolean = parent.exists(_.isArray)
+}
 
+object TypeParser {
   private val logger: Logger = LoggerFactory.getLogger(this.getClass)
-  private val epochTimestampFormat: String = Defaults.getGlobalFormat(TimestampType)
-  private val epochDateFormat: String = Defaults.getGlobalFormat(DateType)
 
-  /** Defines the Spark SQL (DSL) logic for parsing different (primitive) types of columns  */
-  private def primitiveCastLogic(field: StructField, origSchema: StructType, path: String, attr: Column): Column = {
-    field.dataType match {
-      case _: TimestampType => toTimestamp(field, origSchema, path, attr)
-      case _: DateType      => toDate(field, origSchema, path, attr)
-      case t: DataType      => attr.cast(t)
-      case t                => throw new IllegalStateException(s"${t.typeName} type cannot be cast to a strong typed format in this version of Enceladus")
-    }
+  def standardize(field: StructField, path: String, origSchema: StructType): ParseOutput = {
+    TypeParser(field, path, origSchema).standardize()
   }
 
-  /**
-    * Deals with specific weirdness that we see with dates & times coming from various tech stacks
-    * Also supports epoch format
-    */
-  private def dateTimePreProcess(field: StructField, origSchema: StructType, path: String, attr: Column, epochFormat: String, castFunction: (Column, String) => Column): Column = {
-    val format = DateTimePattern(field)
-
-    if (DateTimePattern.isEpoch(format)) {
-      from_unixtime(attr.cast("Long") / format.epochFactor, epochFormat)
-    } else {
-      val origType: Option[DataType] = SchemaUtils.getFieldType(path, origSchema) // sadly with parquet support, incoming might not be all `plain`
-      // this case covers some IBM date format where it's represented as a double ddmmyyyy.hhmmss
-      val sourceCol: Column = origType.get match {
-        case _: DoubleType | _: FloatType =>
-          logger.warn(s"$path is specified as timestamp or date, but original type is Double. Trying to interpret as string.")
-          if (format.isDefault) {
-            throw new InvalidParameterException("Dates & times represented as numeric values need specified 'pattern' metadata")
-          }
-          val Array(l, r) = format.split("\\.")
-          val precision = l.length() + r.length()
-          val scale = r.length()
-          attr.cast(DecimalType.apply(precision, scale)).cast(StringType)
-        case _: DecimalType =>
-          logger.warn(s"$path is specified as timestamp or date, but original type is Decimal. Trying to interpret as string.")
-          if (format.isDefault) {
-            throw new InvalidParameterException("Dates & times represented as numeric values need specified 'pattern' metadata")
-          }
-          attr.cast(StringType)
-        case _ => attr
-      }
-      castFunction(sourceCol, format)
-    }
+  sealed trait Parent {
+    val parentColumn: Column
+    val isArray: Boolean
```

Comment:
A very minor suggestion. This boolean seems just a convenience for distinguishing between `ArrayParent` and `StructParent` since it i always true for `ArrayParent` and always false for `StructParent`.

Consider removing the boolean and replacing its currant usage with pattern matching (or `isInstanceOf()`). Again, it is only a suggestion.